compiler/src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -2080,11 +2098,73 @@ object Parsers {
 
     /** ObjectDef       ::= id TemplateOpt
      */
-    def objectDef(start: Offset, mods: Modifiers, docstring: Option[Comment] = None): ModuleDef = atPos(start, nameStart) {
-      val name = ident()
+    def objectDef(start: Offset, mods: Modifiers): ModuleDef = atPos(start, nameStart) {
+      objectDefRest(start, mods, ident())
+    }
+
+    def objectDefRest(start: Offset, mods: Modifiers, name: TermName): ModuleDef = {
       val template = templateOpt(emptyConstructor)
+      ModuleDef(name, template).withMods(mods).setComment(in.getDocComment(start))
+    }
+
+    /**  id ClassConstr [`extends' [ConstrApps]]
+     *   [nl] ‘{’ EnumCaseStats ‘}’
+     */
+    def enumDef(start: Offset, mods: Modifiers, enumMod: Mod): Thicket = {
+      val point = nameStart
+      val modName = ident()
+      val clsName = modName.toTypeName
+      val constr = classConstr(clsName)
+      val parents =
+        if (in.token == EXTENDS) {
+          in.nextToken();
+          newLineOptWhenFollowedBy(LBRACE)
+          if (in.token == LBRACE) Nil else tokenSeparated(WITH, constrApp)
+        }
+        else Nil
+      val clsDef = atPos(start, point) {
+        TypeDef(clsName, Template(constr, parents, EmptyValDef, Nil))
+          .withMods(addMod(mods, enumMod)).setComment(in.getDocComment(start))
+      }
+      newLineOptWhenFollowedBy(LBRACE)
+      val modDef = atPos(in.offset) {
+        val body = inBraces(enumCaseStats)
+        ModuleDef(modName, Template(emptyConstructor, Nil, EmptyValDef, body))
+          .withMods(mods)
+      }
+      Thicket(clsDef :: modDef :: Nil)
```

Comment:
I tried multi-level expansion first, including flattening, but it did not work. Essentially the compiler got confused with symbol attachments and I did not find a good way around it.